### PR TITLE
Fix/put on busy shards

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

**Comment by carpawell** [comment]
Well, it works. There is not much we can do about testing currently. It is hard to test in general because it is time-dependent. But it is quite straightforward code, and the tests are okay.

**Comment by carpawell** [comment]
Full run: https://rest.fs.neo.org/HXSaMJXk2g8C14ht8HSi7BBaiYZ1HeWh2xnWPGQCg4H6/1997-1729097220/index.html

File: pkg/local_object_storage/engine/put.go [hunk 1]
```diff
@@ -76,38 +79,77 @@ func (e *StorageEngine) put(prm PutPrm) (PutRes, error) {
...
+	return PutRes{}, nil
+}
+
+func (e *StorageEngine) putToShardWithDeadLine(sh hashedShard, ind int, pool util.WorkerPool, addr oid.Address, prm PutPrm) error {
+	const deadline = 30 * time.Second
+	timer := time.NewTimer(deadline)
+	defer timer.Stop()
+
+	const putCooldown = 100 * time.Millisecond
+	ticker := time.NewTicker(putCooldown)
```
Comment (roman-khimov) [inline pkg/local_object_storage/engine/put.go:124]: One of the reasons I don't like pools we have now. Channels provide natural queueing and blocking/unblocking.

**Review by roman-khimov — COMMENTED** [review]
It's an improvement, but channels/routines would be even better.

**Comment by carpawell** [comment]
> but channels/routines would be even better

@roman-khimov, what do you mean by that? Sending tasks via channels to every shard? What is the difference?

**Comment by roman-khimov** [comment]
How can I get a "block for 30s" behavior with a pool? How can I get a queue with a pool? That's the difference.

**Comment by roman-khimov** [comment]
`MaxBlockingTasks` allows to have some queue, maybe we can expose it through the config and test the behavior this way? I think we can have active threads with bigger queues at shard level (although it depends SSD/HDD) instead of current huge pools and no queue. Time to poll-block can be tuned too, we can make additional tests to see how it affects operation.

File: pkg/local_object_storage/engine/put.go [hunk 2]
```diff
@@ -76,38 +79,77 @@ func (e *StorageEngine) put(prm PutPrm) (PutRes, error) {
...
+			return PutRes{}, errPutShard
+		}
 	}
 
-	return PutRes{}, err
+	return PutRes{}, nil
+}
+
+func (e *StorageEngine) putToShardWithDeadLine(sh hashedShard, ind int, pool util.WorkerPool, addr oid.Address, prm PutPrm) error {
+	const deadline = 30 * time.Second
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/engine/put.go]: do you mean node's config? one more value that no one knows what to pick?

File: pkg/local_object_storage/engine/put.go [hunk 3]
```diff
@@ -167,19 +214,20 @@ func (e *StorageEngine) putToShard(sh hashedShard, ind int, pool util.WorkerPool
 				return
 			}
 
-			e.reportShardError(sh, "could not put object to shard", err)
+			e.reportShardError(sh, "could not put object to shard", errGlobal)
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/engine/put.go]: outdated

File: pkg/local_object_storage/engine/put.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/local_object_storage/engine/put.go]: Yeah. This directly affects the reply latency. The other option of course is to check if we have any other timeout value that can be reused.

File: pkg/local_object_storage/engine/put.go [hunk 4]
```diff
@@ -140,10 +157,37 @@ func (e *StorageEngine) putToShard(sh shardWrapper, ind int, pool util.WorkerPoo
...
 	}
 
 	<-exitCh
 
-	return putSuccess, alreadyExists
+	return putSuccess, alreadyExists, overloaded
+}
+
+func (e *StorageEngine) putToShardWithDeadLine(sh shardWrapper, ind int, pool util.WorkerPool, addr oid.Address, obj *objectSDK.Object, objBin []byte, hdrLen int) bool {
+	var deadline = 30 * time.Second
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/engine/put.go]: np

File: pkg/local_object_storage/engine/put.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/engine/put.go]: np

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**